### DISCUSSION TITLE
Consulta sobre rehidratación de los sensores de heladera

### DISCUSSION THREAD

**Discussion by agmblanco** [dds-jv/foro#444] [TPA - 2024] [answered]
Buenas noches. Tenemos algunas dudas sobre la persistencia de los sensores de las heladeras. Entendemos que no deberían ser persistidos pero sí inicializados. Sin embargo, debido a ser componentes externos, llegamos a una limitación en cuando a que al momento de hidratar las instancias de heladera, no contamos con las instancias de los sensores base que deben ser adaptadas y asignadas a la misma. Agradeceríamos si nos pudiesen orientar sobre cómo deberíamos manejar este escenario. 

Desde ya muchas gracias. Saludos, Grupo 5!

**Comment by flbulgarelli** [comment, marked as answer]
:wave: ¡Buenas!

> Entendemos que no deberían ser persistidos pero sí inicializados. 

Correcto. Los sensores son objetos provistos por el API de las heladeras, y hasta donde sabemos no contienen información relevante. 
 
> Sin embargo, debido a ser componentes externos, llegamos a una limitación en cuando a que al momento de hidratar las instancias de heladera, no contamos con las instancias de los sensores base que deben ser adaptadas y asignadas a la misma. 

La respuesta a esta problema depende bastante de qué hayan asumido a la hora de obtener esos componentes. ¿Los instancian a mano? ¿Los obtienen de un singleton? ¿Existe otro componente que los provee? ¿Inyectan sus dependencias? 

En cualquier caso, probablemente la anotación [`@PostLoad`](https://docs.jboss.org/hibernate/jpa/2.1/api/javax/persistence/PostLoad.html) les pueda ser útil, ya sea para instanciar los sensores, localizarlos, o localizar a un objeto que los pueda proveer.

**Reply by agmblanco** [reply]
Hola Franco, muchas gracias por la respuesta! En nuestro caso particular dichas dependencias eran inyectadas. Como en ningún momento tenemos el componente real, por ahora, el único lugar donde se asignan es a la hora de testear; donde utilizando Mockito creamos un mock del objeto "base" del sensor desconocido para luego adaptarlo e inyectarselo a la heladera.

Saludos, grupo 5!

**Reply by flbulgarelli** [reply]
OK! Entonces tienen, a grandes rasgos, tres opciones: 

 1. Utilizar `PostLoad`,  como les comenté arriba, y desde allí utilizar un objeto global que les pueda proveer directa o indirectamente las dependencias que necesitan (un _service locator_, un objeto tipo fábrica, etc). 
 2. Una vez que traen el objeto de la base de datos, manualmente inyectarle (a través de un setter) las mismas dependencias que normalmente le hubieran inyectado por constructor. 
 3. Eliminar esas dependencias del objeto en cuestión y mover ese comportamiento fuera del mismo, o bien mantener ese comportamiento pero parametrizando las dependencias en los métodos en los que las necesitan. 

Cada una de estas estrategias tienen pros y contras, y en última instancia deberán evaluar si es más sencillo introducir ese objeto _localizador_, contar con las dependencias cada vez que se accede al _entity manager_ o contar con esas dependencias cada vez que se usan los métodos que las requieren.
